baseline-network-linux-hld.md

```diff
@@ -0,0 +1,250 @@
+be split into multiple subpackages:
+
+- a base package for the userspace utilities and API libraries
+- one or multiple support packages for the platforms
+
+### ASIC/dataplane support
+
+ASIC/dataplane support is out of scope for this proposal and is intentionally
+left as possible extensions to be implemented by NOS providers (e.g. SONiC or
+DENT) on top of the BSP Layer targeted here.
+
+One exception are switchdev-enabled platforms like Mellanox or Marvell Prestera,
+where the ASIC support is already provided by the upstream Linux kernel.
+
+### Targeted Kernel and Yocto versions
+
+Target Yocto version is scarthgarp, which is the next LTS release, planned for
+April 2024.
+
+The initial Kernel version target could be 5.15 (minimum kernel version of
+Yocto) or 6.6 (the latest LTS version of the Linux kernel). The kernel patch
+version will be a moving target, and regularly updated (ideally weekly).
+
+Since there will be many out-of-tree drivers for the platform support, required
+API updates should be done via
+[coccinelle semantic patches](https://coccinelle.gitlabpages.inria.fr/website/).
+This should ensure that the amount of work needed for keeping drivers building
+and working with newer kernels stays manageable.
+
+The semantic patches will also help vendors quickly updating their drivers when
+submitting new platforms based on older releases.
+
+## Milestones
+
+- M1:
+  - initial platform support repository with hand-selected/-crafted platform
+    support (>= 1 platform)
+  - initial machine definitions
+  - initial layer with kernel with defconfig, and package for platform support
+  - initial kernel 5.15?
+  - ramdisk image bootable on selected platform(s)
+- M2:
+  - refined kernel configuration via kernel-meta
+  - import of various platforms via scripts from SONiC
+  - initial ONIE NOS installer
+- M3:
+  - import missing platforms from ONL (convert them to PDDF?)
+  - import missing platforms from DentOS (convert them to PDDF?)
+  - kernel 6.6 support(?)
+  - coccinelle semantic patches for upgrading drivers to support Linux 6.6 APIs
+  - ONLP adapter(?)
+  - extensible ONIE NOS installer
+- M4:
+  - get accepted as
+    [official yocto layer](https://www.yoctoproject.org/development/yocto-project-compatible-layers/)
+  - are we fine with LTS kernels, or do we need to support non-LTS ones?
+  - figure out long-term maintenance strategy
+
+## Possible Future Extensions
+
+### ASIC / dataplane support
+
+While providing ASIC / dataplane support is out of scope for this project,
+a follow up project could provide appropriate SAI or similar packaging.
+
+### Common userspace API / utilities
+
+Currently there are several competing ways of providing access to sensor data
+and other values. To avoid forcing users to know which one is used for their
+current platform, it would be nice to have a generic/common utilities for
+querying the data, hiding the implementation details.
```

Comment:
It would be useful to list (and perhaps link to) the competing ways here.

Reply:
Good point, we should explicitly list them (ONLP, SONiC PDDF, S3IP) and maybe even explain their differences somehwere.

Maybe even provide some in-kernel interfaces or "common" userspace daemons that implement similar functionality as well (which frankly should be where the effort should go instead of adding yet another abstraction layer ...)

Reply:
The standard kernel interface for this is hwmon. For iio sensors there is an in-kernel translation layer.

Reply:
Right. ONLP etc are often about providing semantic information to the existing sensors (e.g. where in the chassis they are).

But for some components there are no (usable) upstream kernel interfaces yet. E.g. SFP ports (AFAICT) rely on being tied to ethernet interfaces, which don't exist for those devices where the whole ASIC access is hidden behind a binary blob interface.

And there is no way to describe SFP ports outside of devicetree.

Or the PSUs that provide multiple different sensors and EEPROM access via I²C, but their presence is instead described in a register in a separate CPLD.

Also ONIE enabled devices have an I²C connected EEPROM with contains various TLV entries describing the device (model name, serial number, mac addresses etc). To read this out you will need to know which I²C bus to access, at which address. And there is no way to autodetect that.

So these utilities also provide system setup at boot by instantiating the appropriate I²C devices, so things become visible.

Yes there is a certain catch-22 - you need to know the device you are booting on to know on which I²C bus and address the EEPROM is to know which device you are booting on.

Obviously this problem doesn't exist on devicetree devices, but it exists on the many x86 based platforms.

Reply:
Added a list of competing implementations with links. Also tried to summarize their differences a bit.